client/X11/xf_cliprdr.c

```diff
@@ -1663,6 +1928,334 @@ static UINT xf_cliprdr_clipboard_file_range_failure(wClipboardDelegate* delegate
+{
+	double timeout = 0;
+	struct stat stbuf;
+
+	xfCliprdrFuseInode* node;
+	xfClipboard* clipboard = (xfClipboard*)fuse_req_userdata(req);
+
+	ArrayList_Lock(clipboard->ino_list);
+	node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, ino - 1);
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!node)
+	{
+		fuse_reply_err(req, ENOENT);
+	}
+	else
+	{
+		memset(&stbuf, 0, sizeof(stbuf));
+		stbuf.st_ino = ino;
+		stbuf.st_mode = node->st_mode;
+		stbuf.st_mtime = node->st_mtim.tv_sec;
+		if (ino == FUSE_ROOT_ID)
+		{
+			stbuf.st_nlink = 2;
+		}
+		else
+		{
+			stbuf.st_nlink = 1;
+		}
+		fuse_reply_attr(req, &stbuf, timeout);
+	}
+}
+
+static void xf_cliprdr_fuse_readdir(fuse_req_t req, fuse_ino_t ino, size_t size, off_t off,
+                                    struct fuse_file_info* fi)
+{
+	size_t count;
+	size_t index;
+	size_t child_ino;
+	size_t direntry_len;
+	char* buf = (char*)calloc(size, sizeof(char));
+	size_t pos = 0;
+	xfCliprdrFuseInode* child_node;
+	xfCliprdrFuseInode* node;
+	xfClipboard* clipboard = (xfClipboard*)fuse_req_userdata(req);
+
+	ArrayList_Lock(clipboard->ino_list);
+	node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, ino - 1);
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!node || !node->child_inos)
+	{
+		fuse_reply_err(req, ENOENT);
+		return;
+	}
+	else if ((node->st_mode & S_IFDIR) == 0)
+	{
+		fuse_reply_err(req, ENOTDIR);
+		return;
+	}
+
+	ArrayList_Lock(node->child_inos);
+	count = ArrayList_Count(node->child_inos);
+	if (count == 0 || count + 1 <= off)
+	{
+		fuse_reply_buf(req, NULL, 0);
+	}
+	else
+	{
+		for (index = off; index < count + 2; index++)
+		{
+			if (index == 0)
+			{
+				struct stat stbuf;
+				stbuf.st_ino = ino;
+				direntry_len = fuse_add_direntry(req, NULL, 0, ".", NULL, index);
+				if (pos + direntry_len > size)
+				{
+					break;
+				}
+				fuse_add_direntry(req, buf + pos, size - pos, ".", &stbuf, index);
+				pos += direntry_len;
+			}
+			else if (index == 1)
+			{
+				struct stat stbuf;
+				stbuf.st_ino = node->parent_ino;
+				direntry_len = fuse_add_direntry(req, NULL, 0, "..", NULL, index);
+				if (pos + direntry_len > size)
+				{
+					break;
+				}
+				fuse_add_direntry(req, buf + pos, size - pos, "..", &stbuf, index);
+				pos += direntry_len;
+			}
+			else
+			{
+				child_ino = (size_t)ArrayList_GetItem(node->child_inos, index - 2);
+				ArrayList_Lock(clipboard->ino_list);
+				child_node =
+				    (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, child_ino - 1);
+				ArrayList_Unlock(clipboard->ino_list);
+				// child_node must exists?
+				direntry_len = fuse_add_direntry(req, NULL, 0, child_node->name, NULL, index);
+				if (pos + direntry_len > size)
+				{
+					break;
+				}
+
+				struct stat stbuf;
+				stbuf.st_ino = child_node->ino;
+				fuse_add_direntry(req, buf + pos, size - pos, child_node->name, &stbuf, index);
+				pos += direntry_len;
+			}
+		}
+
+		fuse_reply_buf(req, buf, pos);
+		free(buf);
```

Comment:
I guess this better be done after the lock has been released. Who knows how long a FUSE call will take.